File: 2021/02.c
#include <stdio.h>

int main() {
  FILE *input = fopen("input", "r");
  int pos = 0, depth = 0, aim = 0, pos2 = 0, depth2 = 0;
  char action[8];
  int amount;
  while (fscanf(input, "%s %d\n", action, &amount) != EOF) {
    switch (action[0]) {
    case 'f':
      pos += amount;
      pos2 += amount;
      depth2 += aim * amount;
      break;
    case 'd':
      depth += amount;
      aim += amount;
      break;
    case 'u':
      depth -= amount;
      aim -= amount;
      break;
    }
  }
  printf("Part One: %d\nPart Two: %d\n", pos * depth, pos2 * depth2);
  fclose(input);
}

File: src/2021/02.rs
aoc::solution!(2021, 2);

pub fn part_one(input: &str) -> Option<u64> {
    let (pos, depth) = command(input).fold((0, 0), |(pos, depth), (action, amount)| match action {
        b'f' => (pos + amount, depth),
        b'd' => (pos, depth + amount),
        b'u' => (pos, depth - amount),
        _ => unreachable!("there are only three commands"),
    });
    Some(pos * depth)
}

pub fn part_two(input: &str) -> Option<u64> {
    let (pos, depth, _) =
        command(input).fold(
            (0, 0, 0),
            |(pos, depth, aim), (action, amount)| match action {
                b'f' => (pos + amount, depth + aim * amount, aim),
                b'd' => (pos, depth, aim + amount),
                b'u' => (pos, depth, aim - amount),
                _ => unreachable!("there are only three commands"),
            },
        );
    Some(pos * depth)
}

fn command(input: &str) -> impl Iterator<Item = (&u8, u64)> {
    input.lines().map(|line| {
        let (action, amount) = line
            .trim_end()
            .split_once(' ')
            .expect("each line contain an action and a number separated by a space");
        (
            action.as_bytes().first().unwrap(),
            amount.parse().expect("small positive integer"),
        )
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_part_one() {
        let result = part_one(&read_file("examples", YEAR, DAY, Part::Both));
        assert_eq!(result, Some(150));
    }

    #[test]
    fn test_part_two() {
        let result = part_two(&read_file("examples", YEAR, DAY, Part::Both));
        assert_eq!(result, Some(900));
    }
}
